File: backend/app/api/league.py
"""Weekly league endpoints (CR004, D-060).

GET   /v1/league/standings     my cohort's board this week (404 if unassigned)
GET   /v1/league/me            handle, tier, reputation, week points, rank, streak
GET   /v1/league/history       my past weeks (tier, final rank, outcome)
PATCH /v1/league/handle        regenerate the pseudonymous handle (once, ever)
GET   /v1/league/badges        CR091/CR092 — every badge earned, incl. permanent flair
POST  /v1/league/streak/freeze CR094 — consume one of 2 Floor-Manager freezes/year

Reputation-based only — no P&L appears anywhere in this surface (D-060,
store declaration). Handles keep the leaderboard pseudonymous; real names
render only for members who opted in via `users.show_display_name`.
"""

from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel
from sqlalchemy import select

from app.api.dependencies import get_current_user
from app.db import get_session
from app.db.models import User
from app.services.league_service import (
    HandleAlreadyRegenerated,
    get_league_service,
)
from app.services.reputation_service import get_reputation_service


router = APIRouter(prefix="/v1/league", tags=["league"])


def _attached_user(session, user_id) -> User:
    user = session.execute(
        select(User).where(User.id == user_id)
    ).scalar_one_or_none()
    if user is None:  # pragma: no cover — bearer already validated
        raise HTTPException(status.HTTP_404_NOT_FOUND, "user not found")
    return user


@router.get("/standings")
def standings(current_user: User = Depends(get_current_user)) -> dict:
    svc = get_league_service()
    with get_session() as s:
        board = svc.standings(s, current_user.id)
    if board is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "not_in_league")
    return board


@router.get("/me")
def me(current_user: User = Depends(get_current_user)) -> dict:
    league = get_league_service()
    rep = get_reputation_service()
    with get_session() as s:
        user = _attached_user(s, current_user.id)
        handle = league.ensure_handle(s, user)
        member = league.membership(s, user.id)
        rank = None
        tier = league._current_tier(s, user.id)
        points_this_week = 0
        if member is not None:
            points_this_week = member.points
            board = league.standings(s, user.id)
            if board is not None:
                tier = board["tier"]
                rank = next(
                    (m["rank"] for m in board["members"] if m["is_me"]), None,
                )
        streak = rep.streak(s, user.id)
        return {
            "handle": handle,
            "tier": tier,
            "reputation": user.reputation,
            "points_this_week": points_this_week,
            "rank": rank,
            "show_display_name": user.show_display_name,
            "streak": {
                "current": streak.current,
                "longest": streak.longest,
                "next_milestone": streak.next_milestone,
            },
        }


@router.get("/history")
def history(current_user: User = Depends(get_current_user)) -> list[dict]:
    with get_session() as s:
        return get_league_service().history(s, current_user.id)


@router.patch("/handle")
def regenerate_handle(
    current_user: User = Depends(get_current_user),
) -> dict:
    svc = get_league_service()
    with get_session() as s:
        user = _attached_user(s, current_user.id)
        try:
            handle = svc.regenerate_handle(s, user)
        except HandleAlreadyRegenerated:
            raise HTTPException(status.HTTP_409_CONFLICT, "already_regenerated")
    return {"handle": handle}


@router.get("/badges")
def badges(current_user: User = Depends(get_current_user)) -> list[dict]:
    """CR091 read endpoint + CR092 flair exposure. Mobile display of these
    is a separate, unlaned surface (backend contract only)."""
    rep = get_reputation_service()
    with get_session() as s:
        return [
            {
                "badge_key": b.badge_key,
                "earned_at": b.earned_at.isoformat(),
                "is_permanent_flair": b.is_permanent_flair,
            }
            for b in rep.badges(s, current_user.id)
        ]


class StreakFreezeRequest(BaseModel):
    on: str | None = None  # YYYY-MM-DD, defaults to the user's local today


_FREEZE_REASON_STATUS = {
    "not_entitled": status.HTTP_403_FORBIDDEN,
    "limit_reached": status.HTTP_409_CONFLICT,
    "already_frozen": status.HTTP_409_CONFLICT,
}


@router.post("/streak/freeze")
def freeze_streak(
    req: StreakFreezeRequest,
    current_user: User = Depends(get_current_user),
) -> dict:
    """CR094 — Floor Manager only, 2/year. Refuses visibly (D5): a
    non-entitled or exhausted caller gets an explicit HTTP error, never a
    silent 200 that looks like it worked."""
    from datetime import date as date_cls

    on = None
    if req.on is not None:
        try:
            on = date_cls.fromisoformat(req.on)
        except ValueError:
            raise HTTPException(status.HTTP_400_BAD_REQUEST, "on must be YYYY-MM-DD")

    rep = get_reputation_service()
    with get_session() as s:
        result = rep.freeze(s, current_user.id, on=on)
    if not result.ok:
        raise HTTPException(
            _FREEZE_REASON_STATUS.get(result.reason, status.HTTP_400_BAD_REQUEST),
            result.reason,
        )
    return {"ok": True, "remaining": result.remaining}

File: mobile/lib/models/league.dart
/// CR010 (B2/B5) — league + streak models mirroring the untyped `/v1/league/*`
/// JSON. The backend serialises raw dicts (no response_model), so keys are
/// snake_case and there are no aliases. Nullable-tolerant parsing throughout.
library;

class StreakInfo {
  const StreakInfo({
    required this.current,
    required this.longest,
    this.nextMilestone,
  });

  final int current;
  final int longest;
  final int? nextMilestone;

  factory StreakInfo.fromJson(Map<String, dynamic> j) => StreakInfo(
        current: (j['current'] as num?)?.toInt() ?? 0,
        longest: (j['longest'] as num?)?.toInt() ?? 0,
        nextMilestone: (j['next_milestone'] as num?)?.toInt(),
      );
}

/// `GET /v1/league/me` — the caller's league standing + streak. The route
/// mints a pseudonymous handle on first call, so `handle` is always present.
class LeagueMe {
  const LeagueMe({
    required this.handle,
    required this.tier,
    required this.reputation,
    required this.pointsThisWeek,
    required this.rank,
    required this.showDisplayName,
    required this.streak,
  });

  final String handle;
  final String tier;
  final int reputation;
  final int pointsThisWeek;
  final int? rank; // null when unassigned this week
  final bool showDisplayName;
  final StreakInfo streak;

  factory LeagueMe.fromJson(Map<String, dynamic> j) => LeagueMe(
        handle: j['handle'] as String? ?? '—',
        tier: j['tier'] as String? ?? 'apprentice',
        reputation: (j['reputation'] as num?)?.toInt() ?? 0,
        pointsThisWeek: (j['points_this_week'] as num?)?.toInt() ?? 0,
        rank: (j['rank'] as num?)?.toInt(),
        showDisplayName: j['show_display_name'] as bool? ?? false,
        streak: StreakInfo.fromJson(
          (j['streak'] as Map<String, dynamic>?) ?? const {},
        ),
      );
}

class LeagueMemberRow {
  const LeagueMemberRow({
    required this.rank,
    required this.userId,
    required this.handle,
    required this.displayName,
    required this.points,
    required this.isMe,
  });

  final int rank;
  final String userId;
  final String handle;
  final String? displayName; // non-null only when that member opted in
  final int points;
  final bool isMe;

  factory LeagueMemberRow.fromJson(Map<String, dynamic> j) => LeagueMemberRow(
        rank: (j['rank'] as num?)?.toInt() ?? 0,
        userId: j['user_id'] as String? ?? '',
        handle: j['handle'] as String? ?? '—',
        displayName: j['display_name'] as String?,
        points: (j['points'] as num?)?.toInt() ?? 0,
        isMe: j['is_me'] as bool? ?? false,
      );
}

/// `GET /v1/league/standings` — the caller's current-week cohort board.
class LeagueStandings {
  const LeagueStandings({
    required this.leagueId,
    required this.week,
    required this.tier,
    required this.endsAt,
    required this.members,
  });

  final String leagueId;
  final String week;
  final String tier;
  final String endsAt; // ISO 8601
  final List<LeagueMemberRow> members;

  factory LeagueStandings.fromJson(Map<String, dynamic> j) => LeagueStandings(
        leagueId: j['league_id'] as String? ?? '',
        week: j['week'] as String? ?? '',
        tier: j['tier'] as String? ?? 'apprentice',
        endsAt: j['ends_at'] as String? ?? '',
        members: ((j['members'] as List<dynamic>?) ?? const [])
            .map((e) => LeagueMemberRow.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

/// `GET /v1/league/history` — one row per finalised past week.
class LeagueHistoryEntry {
  const LeagueHistoryEntry({
    required this.week,
    required this.tier,
    required this.rankFinal,
    required this.outcome,
    required this.points,
  });

  final String week;
  final String tier;
  final int rankFinal;
  final String? outcome; // "promoted" | "relegated" | "stay"
  final int points;

  factory LeagueHistoryEntry.fromJson(Map<String, dynamic> j) =>
      LeagueHistoryEntry(
        week: j['week'] as String? ?? '',
        tier: j['tier'] as String? ?? 'apprentice',
        rankFinal: (j['rank_final'] as num?)?.toInt() ?? 0,
        outcome: j['outcome'] as String?,
        points: (j['points'] as num?)?.toInt() ?? 0,
      );
}
